### DISCUSSION TITLE
How to import a specific column from a CSV file via COPY when there is no header line?

### DISCUSSION THREAD

**Discussion by parvizfarnia** [duckdb/duckdb#6979] [Q&A]
Hello everyone

I would like to ask a question about the COPY command to import CSV files into DuckDB tables. According to the online documentation here is what we can read in the chapter : [CSV Import/Export](https://duckdb.org/docs/csv_import.html)

> ...  **The amount of columns inside the file must match the amount of columns in the table table_name**, and the contents of the columns must be convertible to the column types of the table. In case this is not possible, an error will be thrown ...
> 
> **If a list of columns is specified, COPY will only copy the data in the specified columns from the file**. If there are any columns in the table that are not in the column list, COPY ... FROM will insert the default values for those columns
> 

Now imagain I have the following table in my database: 

`create table usrs(usrname text);`

And a CSV file with multiple columns where the second column includes usernames that I want to import into my table. In other words, I am interested in only the second column of my CSV file and all other columns are simply to be ignored during the import as if there were never in the file. 

My questions: 

- Is it possible to pick only a specific column from the file and ignore others via COPY? This is possible for example in R using read.csv function.

- If the answer to the above question is yes, how to do so when there is no header line in the file? I mean, my problem is that I don't have a header line but I know that the second column in my CSV file is what I require. Is there any way to make DuckDB understand to pick the second column when there is no header line?

Thanks in advance.

**Comment by Mytherin** [comment]
If there is no header the columns are named `column0, column1, etc...` so you should be able to use those names to select the columns that you want to import.

**Reply by parvizfarnia** [reply]
But according to the syntax of the `COPY` command you can only specify a list of columns based on the available columns in the table, I don't see any option for specifying a column list based on the available columns in the file.

Consider the following example:

`create table usrs(usrname text);`

and the following CSV file where the second column is what I want to import and as you can see there is no header line in the file:

**data.csv**
```
2023-01-01|user1|France
2023-01-14|user2|France
2023-02-17|user3|UK
2023-03-06|user4|USA
2023-03-11|user5|USA
```

How do you write the COPY in order to put (user1, user2, user3, user4, user5) into the above table?

**Reply by Mytherin** [reply]
This is indeed not possible using the `COPY` command. You need to use `read_csv(_auto)`.

**Reply by parvizfarnia** [reply]
Thanks! This is exactly what I was looking for.

```
insert into usrnames(usrname) 
select usrname from read_csv_auto('data.txt', names = ['datavalue', 'usrname', 'country']);
```

Last question: for big CSV files, is `read_csv_auto` as fast as `COPY`?

**Reply by orzom411** [reply]
As a side bar, curiosity with the copy method.  That is I'm curious about this one as I would have thought to copy through a view (a technique I employ frequently, kind of curious why I hadn't tried this here yet, anyway, doesn't work as a view isn't a table).

```
drop table if exists usrs;
create table usrs(usrname text);

drop view if exists vw_usrs;
create view vw_usrs as select usrname as column1 from usrs;
  
COPY vw_usrs FROM 'data.txt' ( DELIMITER '|' );
```

> Error: Catalog Error: vw_usrs is not an table

I would generally suggest the value of views to mask underlying structures, particularly when wrestling with data importation (venders, and clients, seem to take pride in adjusting headers).

Anyway, I was also curious about the initial concept, to copy for a single column blindly (how I read this thread, I think incorrectly read in my part).

`COPY usrs FROM 'data_wh.txt' ( DELIMITER '|' );`

> Error: Invalid Input Error: Error in file "data.txt", on line 1: expected 1 values per row, but got more. (  file=data_wh.txt
>   delimiter='|'
>   quote='"' (default)
>   escape='"' (default)
>   header=0' (default)
>   sample_size=20480
>   ignore_errors=0
>   all_varchar=0)


Sample Data.txt:
2023-01-01|user1|France
2023-01-14|user2|France
2023-02-17|user3|UK
2023-03-06|user4|USA
2023-03-11|user5|USA

Sample Data_wh.txt:
dt|usrname|nation
2023-01-01|user1|France
2023-01-14|user2|France
2023-02-17|user3|UK
2023-03-06|user4|USA
2023-03-11|user5|USA

DuckDB Version: CLI - v0.7.2-dev14

**Reply by Mytherin** [reply]
> Thanks! This is exactly what I was looking for.
> 
> ```
> insert into usrnames(usrname) 
> select usrname from read_csv_auto('data.txt', names = ['datavalue', 'usrname', 'country']);
> ```
> 
> Last question: for big CSV files, is `read_csv_auto` as fast as `COPY`?

Yes, they are exactly the same, except that COPY uses the data types from the table instead of trying to determine the types from the file.

**Reply by parvizfarnia** [reply]
> > Thanks! This is exactly what I was looking for.
> > ```
> > insert into usrnames(usrname) 
> > select usrname from read_csv_auto('data.txt', names = ['datavalue', 'usrname', 'country']);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Last question: for big CSV files, is `read_csv_auto` as fast as `COPY`?
> 
> Yes, they are exactly the same, except that COPY uses the data types from the table instead of trying to determine the types from the file.

Thank you very much for your time and your help.
